File: study/2023/23.02.07/HJH/징검다리.js
function solution(distance, rocks, n) {
  let answer = 0;
  let min = 1;
  let max = distance;
  rocks.sort((a,b) => a-b);
  
  while(min <= max) {
    const mid = Math.floor((min+max) / 2);
    let remove = 0;
    let prev = 0;
    
    for(let i = 0; i < rocks.length; i++) {
      if(rocks[i] - prev <= mid) {
        remove++;
      } else {
        prev = rocks[i];
      }
    }
    
    if(remove > n) {
      max = mid-1;
    } else {
      min = mid+1;
      answer = Math.max(answer, min);
    }
  }
  
  return answer;
}

File: study/2023/23.02.07/EJH/징검다리.java
import java.util.Arrays;

public class 징검다리 {
    public int solution(int distance, int[] rocks, int n) {
        int answer = 0;
        int left = 0, right = distance;

        Arrays.sort(rocks);
        while(left <= right) {
            int mid = (left + right) / 2;
            int removeRocks = 0;
            int prev = 0;

            for(int i = 0; i < rocks.length; i++) {
                if (rocks[i] - prev < mid)
                    removeRocks++;
                else
                    prev = rocks[i];
            }
            if(distance - rocks[rocks.length-1] < mid)
                removeRocks++;
            if(removeRocks <= n) {
                answer = mid;
                left = mid + 1;
            }
            else
                right = mid - 1;
        }
        return answer;
    }
}
